### ISSUE TITLE
Type mismatch with Gitlab API for /protected-branches

### ISSUE THREAD

**Issue by theodorvulcan** [jdalrymple/gitbeaker#3358]
Hey I noticed a type mismatch with the Gitlab API on the protected-branches endpoint.
In the Gitlab API it seems they return all `access_level` , `user_id` , `group_id` which are essentially 3 types of access. Because a single ExtendedProtectedBranchAccessLevel can only be of one of the 3 types only one will be set out of the 3 the other being returned as null. 
Well in your types you don't expect access_level to be null which it could ~~and also you expect user_id and group_id to be missing which seems can't be the case~~ Sorry I just read that they are not present for non premium users so that makes sense.

[Link to Gitlab docs mentioning this](https://docs.gitlab.com/ee/api/protected_branches.html#:~:text=%7B%0A%20%20%20%20%20%20%22id%22%3A%20%201%2C-,%22access_level%22%3A%20null%2C,-%22user_id%22%3A%20null%2C%0A%20%20%20%20%20%20%22group_id).

https://github.com/jdalrymple/gitbeaker/blob/86724f338e9deee46bcc311a85a56b4c44883207/packages/core/src/resources/ProtectedBranches.ts#L13C4-L13C4

**Comment by theodorvulcan** [comment]
Actually testing out the PATCH to add a new user_id it seems the response contradicts their examples by having both access_level and user_id set...
`{"id":1,"access_level":40,"access_level_description":"Username","user_id":1,"group_id":null}`

Though looking deeper into what I need which is add and also remove users from permission lists I did notice another mismatch in types around this area. 
The way to delete a entry from any of the allowed_to_push/merge/unprotect is to make a PATCH with {"id": 1, "_destroy": true}
Well the method that exposes the patch in your api is edit() and those options expects EditProtectedBranchOptions.allowedTo... typed as Record<string, number>[].

I was thinking it could be nice to have stronger types? 🤷
Like:
```
export interface EditProtectedBranchAllowBase {
  _destroy: boolean;
}

export interface EditProtectedBranchAllowUser extends EditProtectedBranchAllowBase {
  user_id: number;
}

...

export interface EditProtectedBranchOptions {
  ...
  allowedToMerge: EditProtectedBranchAllowUser[] | EditProtectedBranchAllowGroup[] | EditProtectedBranchAllowLevel[];
  ...
}

```

**Comment by jdalrymple** [comment]
Seems to be alot of inconsistencies here, so let me summarise before making changes:

1. This type should allow for null and undefined values on the group_id, user_id.

Old Type:
```ts
export interface ExtendedProtectedBranchAccessLevel {
  access_level: ProtectedBranchAccessLevel;
  access_level_description: string;
  user_id?: number;
  group_id?: number;
}
```
New type
```ts
export interface ExtendedProtectedBranchAccessLevel {
  access_level: ProtectedBranchAccessLevel;
  access_level_description: string;
  user_id?: number | null;
  group_id?: number | null;
}
```

2. Because you can technically set multiple things, no restrictions should be added to narrow down the either or functionality therefore keep that as is

3. Update the edit typing to properly handle deletions

old 
```ts
export type EditProtectedBranchOptions = {
  allowForcePush?: boolean;
  allowedToMerge?: Record<string, number>[];
  allowedToPush?: Record<string, number>[];
  allowedToUnprotect?: Record<string, number>[];
  codeOwnerApprovalRequired?: boolean;
};
```

new
```ts
export type EditProtectedBranchAllow  = {
  _destroy?: boolean;
} & Either3<{
  user_id: number;
}, {
  group_id: number;
}, {
  access_level: number;
}>

export type EditProtectedBranchOptions = {
  allowForcePush?: boolean;
  allowedToMerge?: EditProtectedBranchAllow[];
  allowedToPush?: EditProtectedBranchAllow[];
  allowedToUnprotect?: EditProtectedBranchAllow[];
  codeOwnerApprovalRequired?: boolean;
};
```

**Comment by theodorvulcan** [comment]
yes that is looks right 👍

**Comment by theodorvulcan** [comment]
Ah and it also means that ExtendedProtectedBranchAccessLevel would need to contain `id: number` so that I can pass the id to delete

**Comment by jdalrymple** [comment]
:rocket: Issue was released in [`39.10.0`](https://github.com/jdalrymple/gitbeaker/releases/tag/39.10.0) :rocket:

**Comment by theodorvulcan** [comment]
Hey @jdalrymple! Sorry for coming back to this apparently my code review of what you posted wasn't good enough and got lost in all the ids. 
I mentioned it in the initial message that the PATCH needs to be {"id": 1, "_destroy": true} but then my code snippet went against what I said 🤦 and was defining a type for a "user/group_id": 1  and you probably improved on that.

So the EditProtectedBranchAllow now allows for {"user/group_id": 1, "_destroy": true} but not for "id": 1 which funnily enough doesn't break but doesn't delete anything when calling gitlab.

It seems Gitlab didn't struggle too much to validate the inputs of this API so IMHO the type doesn't need to be tighter 🤷‍♂️ 
Seems you can edit existing access rules like (undocumented)
`{"id": 10, "access_level": 30}` 
You can add all the props to the request and it will remove it as long as destroy is there
`{"id": 10, "user_id": 1, "access_level": 30, "_destroy": true}` 
you can even set both a user and a group which theoretically doesn't make sense as it should be one or the other
`{"id": 10, "user_id": 1, "group_id": 2}`
They even allow for creation of new rules like that
`{"user_id": 1, "group_id": 2}` 

So bottom line I think we could just add an `id?: number` field to EditProtectedBranchAllow to allow for deletion/editing a particular existing rule and call it a day. You think that is fair?

**Comment by jdalrymple** [comment]
According to the docs, only one of those props should be passed at a time. So the types should reflect that. I added the id one in that same fashion. Though i only see it in the `allowed_to_unprotect` section
